README.md
# okxAPI

An unofficial wrapper for [Okx exchange v5 API](https://www.okx.com/docs-v5/en/)

## Installation

You can install the released version of okxAPI from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("okxAPI")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("fanggong/okxAPI")
```

## Requirements

Please refer to [okx my api page](https://www.okx.com/account/my-api) regarding V5 API Key creation.

## Example

### REST API

You can implement all REST API requests in Okx exchange by inheriting the `restAPI` class. For example, to implement [Get currencies](https://www.okx.com/docs-v5/en/#rest-api-funding-get-currencies), you can define the class and call the method as follows:

``` r
myRestAPI <- R6::R6Class(
  inherit = restAPI,
  public = list(
    get_currencies = function(ccy, process = "identity") {
      self$get_result(
        api = "/api/v5/asset/currencies", method = "GET", process = process,
        ccy = ccy
      )
    }
  )
)
tmp <- myRestAPI$new(api_key, secret_key, passphrase)
tmp$get_currencies("BTC")
```

For commonly used interfaces, they have been implemented in this package. The naming convention is as follows: for "/api/v5/AAA/BBB", a new class named `restAPIAAA`, which inherits from `restAPI`, has been defined in this package. The `BBB` method in this new class is used to call the API.

### WebSocket API Public Channel

A `websocketAPIpublic` class based on [websocket package](https://github.com/rstudio/websocket) is used here.

Create new `websocketAPIpublic` object and initiate the connection to the server.

``` r
tmp <- websocketAPIpublic$new()
tmp$connect()
```

Subscribe BTC-USDT-SWAP 5m candlesticks data.

``` r
msg <- list(
  op = "subscribe",
  args = list(
    list(channel = "candle5m", instId = "BTC-USDT-SWAP")
  )
)
msg <- jsonlite::toJSON(msg, auto_unbox = TRUE, pretty = TRUE)
tmp$send(msg)
```

Pass your own callback function

``` r
tmp$on_message(function(event) {
  if (event$data == "pong") {
    cat("Bingo!!")
  }
})
tmp$send("ping")
```

Close the connection.

``` r
tmp$close()
```

### WebSocket API Private Channel

A `websocketAPIprivate` class based on [websocket package](https://github.com/rstudio/websocket) is used here.

Create new `websocketAPIprivate` object and initiate the connection to the server.

``` r
tmp <- websocketAPIprivate$new(api_key, secret_key, passphrase)
tmp$connect()
Sys.sleep(1)  # Waiting for connection success
tmp$login()
```

Subscribe account information.

``` r
msg <- list(
  op = "subscribe",
  args = list(
    list(channel = "account", ccy = "USDT")
  )
)
msg <- jsonlite::toJSON(msg, auto_unbox = TRUE, pretty = TRUE)
tmp$send(msg)
```

Pass your own callback function.

``` r
tmp$on_message(function(event) {
  if (event$data == "pong") {
    cat("Bingo!!")
  }
})
tmp$send("ping")
```

Close the connection.

``` r
tmp$close()
```


NEWS.md
# okxAPI (development version)

# okxAPI 0.1.1

* Added a `NEWS.md` file to track changes to the package.


R/funcs.R
#' @name get_functions
#'
#' @title Wrapper for some frequently used APIs to get data easily
#'
#' @description The main purpose is to handle APIs that have limitations on the number of results returned per single request.
#' @seealso
#' \code{\link{get_positions_history}}
#' \code{\link{get_history_candles}}
NULL


#' @title Retrieve the position data
#'
#' @description Wrapper for API [Get positions history](https://www.okx.com/docs-v5/en/#rest-api-account-get-positions-history).
#'
#' @param api_key Okx API key.
#' @param secret_key Okx API secret key.
#' @param passphrase Okx API passphrase.
#' @param count Retrieve position data for a specified number of past days, with a maximum of 90(days)
#' @param period Due to the 'Number of results per request' limitation of the API,
#' the \code{period} parameter must be specified to ensure that the number of position data entries within each period does not exceed 100.
#' @param ... Other request parameters to be passed, See
#' [Get positions history](https://www.okx.com/docs-v5/en/#rest-api-account-get-positions-history) for more information.
#'
#' @return Position data
#'
#' @examples
#' \dontrun{
#' positions <- get_positions_history(
#'   api_key, secret_key, passphrase, count = 90, period = 10,
#'   instType = "SWAP", mgnMode = "isolated"
#' )
#' }
#'
#' @import data.table
#' @export
get_positions_history <- function(
  api_key, secret_key, passphrase, count = 90, period = 10, ...
) {
  account <- restAPIaccount$new(api_key, secret_key, passphrase)
  now <- time2ts(Sys.time())
  start <- now - count*24*60*60
  end <- start + period*24*60*60 - 0.001
  dat <- list()
  while (start < now) {
    result <- account$positions_history(
      before = as.character(1000*start - 1), after = as.character(1000*end + 1), ...
    )
    if (result$code == "0") {
      dat <- c(dat, result$data)
      message("From ", ts2time(start), " to ", ts2time(end), " complete")
      start <- end + 0.001
      end <- start + period*24*60*60 - 0.001
      Sys.sleep(15)
    }
  }
  dat <- lapply(dat, data.table::as.data.table)
  dat <- data.table::rbindlist(dat)
  to_time <- c("cTime", "uTime")
  dat[, (to_time) := lapply(.SD, ts2time), .SDcols = to_time]
  to_numeric <- c("closeAvgPx", "closeTotalPos", "lever", "openAvgPx", "openMaxPos", "pnl", "pnlRatio")
  dat[, (to_numeric) := lapply(.SD, as.numeric), .SDcols = to_numeric]
  dat
}

#' @title Retrieve the candlestick charts
#'
#' @description Wrapper for API [Get candlesticks](https://www.okx.com/docs-v5/en/#rest-api-market-data-get-candlesticks)
#' and [Get candlesticks history](https://www.okx.com/docs-v5/en/#rest-api-market-data-get-candlesticks-history).
#'
#' @param api_key Okx API key.
#' @param secret_key Okx API secret key.
#' @param passphrase Okx API passphrase.
#' @param bar Bar size, the default is 1m, e.g. 1m/3m/5m/15m/30m/1H/2H/4H, Hong Kong time opening price k-line: 6H/12H/1D/2D/3D.
#' @param count Number of Bars.
#' @param instId Instrument ID, e.g. BTC-USDT-SWAP.
#' @param ... Other request parameters to be passed, See [Get candlesticks history](https://www.okx.com/docs-v5/en/#rest-api-market-data-get-candlesticks-history) for more information.
#'
#' @return Candlestick charts data
#'
#' @examples
#' \dontrun{
#' candles <- get_history_candles(
#'   api_key, secret_key, passphrase, bar = "1m",
#'   count = 24*60, instId = "CFX-USDT-SWAP"
#' )
#' }
#'
#' @import data.table
#' @export
get_history_candles <- function(
  api_key, secret_key, passphrase,
  bar = c("1m", "3m", "5m", "15m", "30m", "1H", "4H", "6H", "12H", "1D", "2D", "3D"),
  count, instId, ...
) {
  bar <- match.arg(bar)
  period <- str2period(bar)

  market <- restAPImarket$new(api_key, secret_key, passphrase)

  now <- time2ts(Sys.time())
  end <- now
  start <- end - ifelse(count >= 100, 100, count) * period + 0.001
  dat <- list()
  for (i in 1:ceiling(count / 100)) {
    if (i == 1) {
      result <- market$candles(
        instId = instId, before = as.character(1000*start - 1),
        after = as.character(1000*end + 1), bar = bar, ...
      )
    } else {
      result <- market$history_candles(
        instId = instId, before = as.character(1000*start - 1),
        after = as.character(1000*end + 1), bar = bar, ...
      )
    }
    if (result$code == "0") {
      dat <- c(dat, result$data)
      message("From ", ts2time(start), " to ", ts2time(end), " complete")
      count <- count - 100
      end <- start - 0.001
      start <- end - ifelse(count >= 100, 100, count) * period + 0.001
      # Sys.sleep(1/20)
    }
  }
  col_names <- c("ts", "open", "high", "low", "close", "vol", "volCcy", "volCcyQuote", "confirm")
  dat <- lapply(dat, as.data.frame, col.names = c(col_names))
  dat <- data.table::rbindlist(dat)
  dat$ts <- ts2time(dat$ts)
  to_numeric <- c("open", "high", "low", "close", "vol", "volCcy", "volCcyQuote")
  dat[, (to_numeric) := lapply(.SD, as.numeric), .SDcols = to_numeric]
  dat
}


R/restAPI.R
#' @title restAPI Class
#'
#' @description Base class for [Okx exchange v5 API](https://www.okx.com/docs-v5/en/).
#'
#' @details You can implement all REST API requests in Okx exchange by inheriting the \code{restAPI} class.
#' Please refer to the example provided at the end of the document.
#'
#' For commonly used interfaces, they have been implemented in this package.
#' The naming convention is as follows: for "/api/v5/AAA/BBB", a new class named \code{restAPIAAA},
#' which inherits from \code{restAPI}, has been defined in this package.
#' The \code{BBB} method in this new class is used to call the API.
#'
#' @examples
#' \dontrun{
#' # for [Get currencies](https://www.okx.com/docs-v5/en/#rest-api-funding-get-currencies)
#' # you can define the class like this
#' myRestAPI <- R6::R6Class(
#'   inherit = restAPI,
#'   public = list(
#'     get_currencies = function(ccy, process = "identity") {
#'       self$get_result(
#'         api = "/api/v5/asset/currencies", method = "GET", process = process,
#'         ccy = ccy
#'       )
#'     }
#'   )
#' )
#' # And call it like this
#' tmp <- myRestAPI$new(api_key, secret_key, passphrase)
#' tmp$get_currencies("BTC")
#' }
#'
#' @seealso
#' \code{\link{restAPItrade}}
#' \code{\link{restAPIaccount}}
#' \code{\link{restAPImarket}}
#'
#' @import R6 httr jsonlite base64enc base64enc
#' @export
restAPI <- R6::R6Class(
  "restAPI",
  public = list(
    #' @field url Okx REST API url, which is https://www.okx.com.
    url = "https://www.okx.com",
    #' @field api_key Okx API key.
    api_key = NA,
    #' @field secret_key Okx API secret key.
    secret_key = NA,
    #' @field passphrase Okx API passphrase.
    passphrase = NA,
    #' @field simulate Whether to use demo trading service.
    simulate = NA,
    #' @description Create a new REST API object.
    #' @param api_key Okx API key.
    #' @param secret_key Okx API secret key.
    #' @param passphrase Okx API passphrase.
    #' @param simulate Whether to use demo trading service.
    initialize = function(api_key, secret_key, passphrase, simulate = FALSE) {
      self$api_key <- api_key
      self$secret_key <- secret_key
      self$passphrase <- passphrase
      self$simulate <- simulate
    },
    #' @description Get UTC timestamp.
    get_timestamp = function() {
      timestamp <- as.numeric(Sys.time()) * 1000
      timestamp <- as.POSIXct(timestamp / 1000, origin = "1970-01-01", tz = "UTC")
      timestamp <- format(timestamp, format = "%Y-%m-%dT%H:%M:%OS3Z", usetz = FALSE)
      timestamp
    },
    #' @description Get request path.
    #' @param api Request api e.g. /api/v5/account/positions-history.
    #' @param method Request method, \code{GET} or \code{POST}.
    #' @param ... Other request parameters.
    get_request_path = function(api, method = c("GET", "POST"), ...) {
      method <- match.arg(method)
      if (method == "GET") {
        tmp <- list(...)
        if (length(tmp) == 0) {
          request_path <- api
        } else {
          tmp <- tmp[!is.na(tmp)]
          request_path <- paste0(api, "?")
          for (i in 1:length(tmp)) {
            tmp[[i]] <- paste0(names(tmp)[i], "=", paste(tmp[[i]], collapse = ","))
          }
          tmp <- paste(unlist(tmp), collapse = "&")
          request_path <- paste0(api, "?", tmp)
        }
      } else {
        request_path <- api
      }
      request_path
    },
    #' @description Get request body.
    #' @param method Request method, \code{GET} or \code{POST}.
    #' @param ... Other request parameters.
    get_body = function(method = c("GET", "POST"), ...) {
      method <- match.arg(method)
      if (method == "GET") {
        body <- ""
      } else {
        tmp <- list(...)
        tmp <- tmp[!is.na(tmp)]
        body <- jsonlite::toJSON(tmp, auto_unbox = TRUE)
      }
      body
    },
    #' @description Get the signing messages.
    #' @param timestamp Retrieve through method \code{get_timestamp}.
    #' @param request_path Retrieve through method \code{get_request_path}.
    #' @param body Retrieve through method \code{get_body}.
    #' @param method Request method, \code{GET} or \code{POST}.
    get_message = function(timestamp, request_path, body, method = c("GET", "POST")) {
      method <- match.arg(method)
      paste0(timestamp, method, request_path, body)
    },
    #' @description Get the signature.
    #' @param msg Retrieve through method \code{get_message}.
    #' @param secret_key Okx API secret key.
    get_signature = function(msg, secret_key = self$secret_key) {
      base64enc::base64encode(digest::hmac(secret_key, msg, algo = "sha256", raw = TRUE))
    },
    #' @description Get request headers.
    #' @param timestamp Retrieve through method \code{get_timestamp}.
    #' @param msg Retrieve through method \code{get_message}.
    get_header = function(timestamp, msg) {
      headers <- c(
        "OK-ACCESS-KEY" = self$api_key,
        "OK-ACCESS-SIGN" = self$get_signature(msg),
        "OK-ACCESS-TIMESTAMP" = timestamp,
        "OK-ACCESS-PASSPHRASE" = self$passphrase,
        "Content-Type" = "application/json"
      )
      if (self$simulate) {
        headers <- c(headers, "x-simulated-trading" = "1")
      }
      headers
    },
    #' @description Retrieve data from api.
    #' @param api Request api e.g. /api/v5/account/positions-history.
    #' @param method Request method, \code{GET} or \code{POST}.
    #' @param process A function to process the data received from the API.
    #' @param ... Other request parameters.
    get_result = function(api, method = c("GET", "POST"), process, ...) {
      method <- match.arg(method)
      timestamp <- self$get_timestamp()
      request_path <- self$get_request_path(api = api, method = method, ...)
      body <- self$get_body(method = method, ...)
      msg <- self$get_message(timestamp, request_path, body, method)
      if (method == "GET") {
        response <- httr::GET(
          paste0(self$url, request_path),
          httr::add_headers(.headers = self$get_header(timestamp, msg))
        )
      } else {
        response <- httr::POST(
          paste0(self$url, request_path),
          httr::add_headers(.headers = self$get_header(timestamp, msg)),
          body = body
        )
      }
      result <- httr::content(response)
      if (httr::http_error(response)) {
        message("Error code ", result$code, ": ", result$msg)
      } else if (result$code != "0") {
        message("Error code ", result$data[[1]]$sCode, ": ", result$data[[1]]$sMsg)
      } else {
        return(do.call(process, list(result)))
      }
    }
  )
)


#' @title restAPItrade Class
#'
#' @description Wrapper for [REST API TRADE](https://www.okx.com/docs-v5/en/#rest-api-trade).
#'
#' @import R6
#' @export
restAPItrade <- R6::R6Class(
  "restAPItrade",
  inherit = restAPI,
  public = list(
    #' @description See [Place order](https://www.okx.com/docs-v5/en/#rest-api-trade-place-order) for more information.
    #' @param instId Instrument ID, e.g. BTC-USD-190927-5000-C.
    #' @param tdMode Trade mode. Margin mode: \code{cross} or \code{isolated.} Non-Margin mode: \code{cash}.
    #' @param side Order side, \code{buy} or \code{sell}.
    #' @param sz Quantity to buy or sell.
    #' @param ordType Order type. \code{market}: Market order, \code{limit}: Limit order, \code{post_only}: Post-only order,
    #' \code{fok}: Fill-or-kill order, \code{ioc}: Immediate-or-cancel order,
    #' \code{optimal_limit_ioc}: Market order with immediate-or-cancel order (applicable only to Futures and Perpetual swap).
    #' @param process A function to process the data received from the API. Default to \code{identity}.
    #' @param ... Other request parameters.
    order  = function(
      instId, tdMode = c("isolated", "cross", "cash"), side = c("buy", "sell"), sz,
      ordType = c("market", "limit", "post_only", "fok", "ioc", "optimal_limit_ioc"),
      process = "identity", ...
    ) {
      tdMode <- match.arg(tdMode)
      side <- match.arg(side)
      ordType <- match.arg(ordType)

      self$get_result(
        api = "/api/v5/trade/order", method = "POST", process = process,
        instId = instId, tdMode = tdMode, side = side, sz = sz, ordType = ordType, ...
      )
    },
    #' @description See [Cancel order](https://www.okx.com/docs-v5/en/#rest-api-trade-cancel-order) for more information.
    #' @param instId Instrument ID, e.g. BTC-USD-190927.
    #' @param ordId Order ID, Either \code{ordId} or \code{clOrdId} is required. If both are passed, \code{ordId} will be used.
    #' @param clOrdId Client Order ID as assigned by the client.
    #' @param process A function to process the data received from the API. Default to \code{identity}.
    #' @param ... Other request parameters.
    cancel_order = function(instId, ordId, clOrdId, process = "identity", ...) {
      if (missing(ordId)) ordId <- NA
      if (missing(clOrdId)) clOrdId <- NA
      self$get_result(
        api = "/api/v5/trade/cancel-order", method = "POST", process = process,
        instId = instId, ordId = ordId, clOrdId = clOrdId, ...
      )
    }
  )
)

#' @title restAPIaccount Class
#'
#' @description Wrapper for [REST API ACCOUNT](https://www.okx.com/docs-v5/en/#rest-api-account).
#'
#' @import R6
#' @export
restAPIaccount <- R6::R6Class(
  "restAPIaccount",
  inherit = restAPI,
  public = list(
    #' @description See [Get balance](https://www.okx.com/docs-v5/en/#rest-api-account-get-balance) for more information.
    #' @param ccy Single currency or a vector composed of multiple currencies. (no more than 20).
    #' @param process A function to process the data received from the API. Default to \code{identity}.
    balance = function(ccy, process = "identity") {
      self$get_result(
        api = "/api/v5/account/balance", method = "GET", process = process,
        ccy = ccy
      )
    },
    #' @description See [Get positions](https://www.okx.com/docs-v5/en/#rest-api-account-get-positions) for more information.
    #' @param process A function to process the data received from the API. Default to \code{identity}.
    #' @param ... Other request parameters.
    positions = function(process = "identity", ...) {
      self$get_result(
        api = "/api/v5/account/positions", method = "GET", process = process,
        ...
      )
    },
    #' @description See [Get positions history](https://www.okx.com/docs-v5/en/#rest-api-account-get-positions-history) for more information.
    #' @param process A function to process the data received from the API. Default to \code{identity}.
    #' @param ... Other request parameters.
    positions_history = function(process = "identity", ...) {
      self$get_result(
        api = "/api/v5/account/positions-history", method = "GET", process = process,
        ...
      )
    }
  )
)


#' @title restAPImarket Class
#'
#' @description Wrapper for [REST API MARKET](https://www.okx.com/docs-v5/en/#rest-api-market-data).
#'
#' @import R6
#' @export
restAPImarket <- R6::R6Class(
  "restAPImarket",
  inherit = restAPI,
  public = list(
    #' @description See [Get order book](https://www.okx.com/docs-v5/en/#rest-api-market-data-get-order-book) for more information.
    #' @param instId Instrument ID, e.g. CFX-USDT
    #' @param process A function to process the data received from the API. Default to \code{identity}.
    #' @param ... Other request parameters.
    books = function(instId, process = "identity", ...) {
      self$get_result(
        api = "/api/v5/market/books", method = "GET", process = process,
        instId = instId, ...
      )
    },
    #' @description See [Get candlesticks](https://www.okx.com/docs-v5/en/#rest-api-market-data-get-candlesticks) for more information.
    #' @param instId Instrument ID, e.g. BTC-USD-190927-5000-C.
    #' @param process A function to process the data received from the API. Default to \code{identity}.
    #' @param ... Other request parameters.
    candles = function(instId, process = "identity", ...) {
      self$get_result(
        api = "/api/v5/market/candles", method = "GET", process = process,
        instId = instId, ...
      )
    },
    #' @description See [Get candlesticks history](https://www.okx.com/docs-v5/en/#rest-api-market-data-get-candlesticks-history) for more information.
    #' @param instId Instrument ID, e.g. BTC-USD-190927-5000-C.
    #' @param process A function to process the data received from the API. Default to \code{identity}.
    #' @param ... Other request parameters.
    history_candles = function(instId, process = "identity", ...) {
      self$get_result(
        api = "/api/v5/market/history-candles", method = "GET", process = process,
        instId = instId, ...
      )
    }
  )
)


R/websocketAPIpublic.R
#' @title websocketAPIpublic Class
#'
#' @description Public channel of WebSocket API for [Okx exchange v5 API](https://www.okx.com/docs-v5/en/).
#' See [Public Channel](https://www.okx.com/docs-v5/en/#websocket-api-public-channel) for more information.
#'
#' @examples
#' \dontrun{
#' tmp <- websocketAPIpublic$new()
#' tmp$connect()
#'
#' # subscribe BTC-USDT-SWAP 5m candlesticks data
#' msg <- list(
#'   op = "unsubscribe",
#'   args = list(
#'     list(channel = "candle5m", instId = "BTC-USDT-SWAP")
#'   )
#' )
#' msg <- jsonlite::toJSON(msg, auto_unbox = TRUE, pretty = TRUE)
#' tmp$send(msg)
#'
#' # pass your own callback function
#' tmp$on_message(function(event) {
#'   if (event$data == "pong") {
#'     cat("Bingo!!\n")
#'   }
#' })
#' tmp$send("ping")
#'
#' tmp$close()
#' }
#'
#' @import R6 websocket
#' @export
websocketAPIpublic <- R6::R6Class(
  "websocketAPIpublic",
  public = list(
    #' @field channel Public WebSocket url.
    channel = "wss://ws.okx.com:8443/ws/v5/public",
    #' @field simulate Whether to use demo trading service.
    simulate = NA,
    #' @field ws A websocket::WebSocket object to establish a connection to the server.
    ws = NA,
    #' @description Create a new websocketAPIpublic object.
    #' @param simulate Whether to use demo trading service.
    initialize = function(simulate = FALSE) {
      self$simulate <- simulate
      if (simulate) self$channel <- "wss://wspap.okx.com:8443/ws/v5/public?brokerId=9999"
      self$ws <- websocket::WebSocket$new(self$channel, autoConnect = FALSE)
    },
    #' @description Initiate the connection to the server.
    connect = function() {
      # self$ws$onOpen(function(event) {
      #   cat("Connection opened\n")
      # })
      #
      # self$ws$onMessage(function(event) {
      #   cat("Client got msg: ", event$data, "\n")
      # })
      #
      # self$ws$onClose(function(event) {
      #   cat("Client disconnected with code ", event$code, ": ", event$reason, "\n")
      # })
      #
      # self$ws$onError(function(event) {
      #   cat("Client failed to connect: ", event$message, "\n")
      # })

      self$ws$connect()
    },
    #' @description Called when the connection is established.
    #' @param func A Callback function.
    on_open = function(func) {
      self$ws$onOpen(func)
    },
    #' @description Called when a previously-opened connection is closed.
    #' The event will have 'code' (integer) and 'reason' (one-element character)
    #' elements that describe the remote's reason for closing.
    #' @param func A Callback function.
    on_close = function(func) {
      self$ws$onClose(func)
    },
    #' @description Called each time a message is received from the server.
    #' The event will have a 'data' element, which is the message content.
    #' @param func A Callback function.
    on_message = function(func) {
      self$ws$onMessage(func)
    },
    #' @description Called when the connection fails to be established.
    #' The event will have an 'message' element, a character vector of length 1
    #' describing the reason for the error.
    #' @param func A Callback function.
    on_error = function(func) {
      self$ws$onError(func)
    },
    #' @description Send a message to the server.
    #' @param msg Messages.
    send = function(msg) {
      self$ws$send(msg)
    },
    #' @description Close the connection.
    close = function() {
      self$ws$close()
    }
  )
)



R/websockerAPIprivate.R
#' @title websocketAPIprivate Class
#'
#' @description Private channel of WebSocket API for [Okx exchange v5 API](https://www.okx.com/docs-v5/en/).
#' See [Private Channel](https://www.okx.com/docs-v5/en/#websocket-api-private-channel) for more information.
#'
#' @examples
#' \dontrun{
#' tmp <- websocketAPIprivate$new(api_key, secret_key, passphrase)
#' tmp$connect()
#' Sys.sleep(1)
#' tmp$login()
#'
#' # subscribe account information
#' msg <- list(
#'   op = "subscribe",
#'   args = list(
#'     list(channel = "account", ccy = "USDT")
#'   )
#' )
#' msg <- jsonlite::toJSON(msg, auto_unbox = TRUE, pretty = TRUE)
#' tmp$send(msg)
#'
#' # pass your own callback function
#' tmp$on_message(function(event) {
#'   if (event$data == "pong") {
#'     cat("Bingo!!\n")
#'   }
#' })
#' tmp$send("ping")
#'
#' tmp$close()
#' }
#'
#' @import R6 websocket base64enc digest
#' @export
websocketAPIprivate <- R6::R6Class(
  "websockerAPIprivate",
  public = list(
    #' @field channel Private WebSocket url.
    channel = "wss://ws.okx.com:8443/ws/v5/private",
    #' @field api_key Okx API key.
    api_key = NA,
    #' @field secret_key Okx API secret key.
    secret_key = NA,
    #' @field passphrase Okx API passphrase.
    passphrase = NA,
    #' @field simulate Whether to use demo trading service.
    simulate = NA,
    #' @field ws A websocket::WebSocket object to establish a connection to the server.
    ws = NA,
    #' @description Craeate a new websocketAPIprivate object.
    #' @param api_key Okx API key.
    #' @param secret_key Okx API secret key.
    #' @param passphrase Okx API passphrase.
    #' @param simulate Whether to use demo trading service.
    initialize = function(api_key, secret_key, passphrase, simulate = FALSE) {
      self$api_key <- api_key
      self$secret_key <- secret_key
      self$passphrase <- passphrase
      self$simulate <- simulate
      if (simulate) self$channel <- "wss://wspap.okx.com:8443/ws/v5/private?brokerId=9999"
      self$ws <- websocket::WebSocket$new(self$channel, autoConnect = FALSE)
    },
    #' @description Get UTC timestamp.
    get_timestamp = function() {
      as.integer(Sys.time())
    },
    #' @description Get the signing messages.
    #' @param timestamp Retrieve through method \code{get_timestamp}.
    get_message = function(timestamp) {
      paste0(timestamp, "GET", "/users/self/verify")
    },
    #' @description Get the signature.
    #' @param secret_key Okx API secret key.
    #' @param msg Retrieve through method \code{get_message}.
    get_signature = function(secret_key, msg) {
      base64enc::base64encode(digest::hmac(
        secret_key, msg, algo = "sha256", raw = TRUE
      ))
    },
    #' @description Initiate the connection to the server.
    connect = function() {
      self$ws$onOpen(function(event) {
        cat("Connection opened\n")
      })

      self$ws$onMessage(function(event) {
        cat("Client got msg: ", event$data, "\n")
      })

      self$ws$onClose(function(event) {
        cat("Client disconnected with code ", event$code, ": ", event$reason, "\n")
      })

      self$ws$onError(function(event) {
        cat("Client failed to connect: ", event$message, "\n")
      })

      self$ws$connect()
    },
    #' @description Log in.
    login = function() {
      status <- attr(self$ws$readyState(), "description") == "Open"
      if (status) {
        timestamp <- self$get_timestamp()
        msg <- self$get_message(timestamp)
        signature <- self$get_signature(self$secret_key, msg)

        login_msg <- list(
          "op" = "login",
          "args" = list(
            list(
              "apiKey" = self$api_key,
              "passphrase" = self$passphrase,
              "timestamp" = timestamp,
              "sign" = signature
            )
          )
        )
        login_msg <- jsonlite::toJSON(login_msg, auto_unbox = TRUE)
        self$ws$send(login_msg)
      } else {
        stop("Connection is not open")
      }
    },
    #' @description Called when the connection is established.
    #' @param func A Callback function.
    on_open = function(func) {
      self$ws$onOpen(func)
    },
    #' @description Called when a previously-opened connection is closed.
    #' The event will have 'code' (integer) and 'reason' (one-element character)
    #' elements that describe the remote's reason for closing.
    #' @param func A Callback function.
    on_close = function(func) {
      self$ws$onClose(func)
    },
    #' @description Called each time a message is received from the server.
    #' The event will have a 'data' element, which is the message content.
    #' @param func A Callback function.
    on_message = function(func) {
      self$ws$onMessage(func)
    },
    #' @description Called when the connection fails to be established.
    #' The event will have an 'message' element, a character vector of length 1
    #' describing the reason for the error.
    #' @param func A Callback function.
    on_error = function(func) {
      self$ws$onError(func)
    },
    #' @description Send a message to the server.
    #' @param msg Messages.
    send = function(msg) {
      self$ws$send(msg)
    },
    #' @description Close the connection.
    close = function() {
      self$ws$close()
    }
  )
)


R/websocketAPIprivate.R
#' @title websocketAPIprivate Class
#'
#' @description Private channel of WebSocket API for [Okx exchange v5 API](https://www.okx.com/docs-v5/en/).
#' See [Private Channel](https://www.okx.com/docs-v5/en/#websocket-api-private-channel) for more information.
#'
#' @examples
#' \dontrun{
#' tmp <- websocketAPIprivate$new(api_key, secret_key, passphrase)
#' tmp$connect()
#' Sys.sleep(1)
#' tmp$login()
#'
#' # subscribe account information
#' msg <- list(
#'   op = "subscribe",
#'   args = list(
#'     list(channel = "account", ccy = "USDT")
#'   )
#' )
#' msg <- jsonlite::toJSON(msg, auto_unbox = TRUE, pretty = TRUE)
#' tmp$send(msg)
#'
#' # pass your own callback function
#' tmp$on_message(function(event) {
#'   if (event$data == "pong") {
#'     cat("Bingo!!\n")
#'   }
#' })
#' tmp$send("ping")
#'
#' tmp$close()
#' }
#'
#' @import R6 websocket base64enc digest
#' @export
websocketAPIprivate <- R6::R6Class(
  "websockerAPIprivate",
  public = list(
    #' @field channel Private WebSocket url.
    channel = "wss://ws.okx.com:8443/ws/v5/private",
    #' @field api_key Okx API key.
    api_key = NA,
    #' @field secret_key Okx API secret key.
    secret_key = NA,
    #' @field passphrase Okx API passphrase.
    passphrase = NA,
    #' @field simulate Whether to use demo trading service.
    simulate = NA,
    #' @field ws A websocket::WebSocket object to establish a connection to the server.
    ws = NA,
    #' @description Craeate a new websocketAPIprivate object.
    #' @param api_key Okx API key.
    #' @param secret_key Okx API secret key.
    #' @param passphrase Okx API passphrase.
    #' @param simulate Whether to use demo trading service.
    initialize = function(api_key, secret_key, passphrase, simulate = FALSE) {
      self$api_key <- api_key
      self$secret_key <- secret_key
      self$passphrase <- passphrase
      self$simulate <- simulate
      if (simulate) self$channel <- "wss://wspap.okx.com:8443/ws/v5/private?brokerId=9999"
      self$ws <- websocket::WebSocket$new(self$channel, autoConnect = FALSE)
    },
    #' @description Get UTC timestamp.
    get_timestamp = function() {
      as.integer(Sys.time())
    },
    #' @description Get the signing messages.
    #' @param timestamp Retrieve through method \code{get_timestamp}.
    get_message = function(timestamp) {
      paste0(timestamp, "GET", "/users/self/verify")
    },
    #' @description Get the signature.
    #' @param secret_key Okx API secret key.
    #' @param msg Retrieve through method \code{get_message}.
    get_signature = function(secret_key, msg) {
      base64enc::base64encode(digest::hmac(
        secret_key, msg, algo = "sha256", raw = TRUE
      ))
    },
    #' @description Initiate the connection to the server.
    connect = function() {
      # self$ws$onOpen(function(event) {
      #   cat("Connection opened\n")
      # })
      #
      # self$ws$onMessage(function(event) {
      #   cat("Client got msg: ", event$data, "\n")
      # })
      #
      # self$ws$onClose(function(event) {
      #   cat("Client disconnected with code ", event$code, ": ", event$reason, "\n")
      # })
      #
      # self$ws$onError(function(event) {
      #   cat("Client failed to connect: ", event$message, "\n")
      # })

      self$ws$connect()
    },
    #' @description Log in.
    login = function() {
      status <- attr(self$ws$readyState(), "description") == "Open"
      if (status) {
        timestamp <- self$get_timestamp()
        msg <- self$get_message(timestamp)
        signature <- self$get_signature(self$secret_key, msg)

        login_msg <- list(
          "op" = "login",
          "args" = list(
            list(
              "apiKey" = self$api_key,
              "passphrase" = self$passphrase,
              "timestamp" = timestamp,
              "sign" = signature
            )
          )
        )
        login_msg <- jsonlite::toJSON(login_msg, auto_unbox = TRUE)
        self$ws$send(login_msg)
      } else {
        stop("Connection is not open")
      }
    },
    #' @description Called when the connection is established.
    #' @param func A Callback function.
    on_open = function(func) {
      self$ws$onOpen(func)
    },
    #' @description Called when a previously-opened connection is closed.
    #' The event will have 'code' (integer) and 'reason' (one-element character)
    #' elements that describe the remote's reason for closing.
    #' @param func A Callback function.
    on_close = function(func) {
      self$ws$onClose(func)
    },
    #' @description Called each time a message is received from the server.
    #' The event will have a 'data' element, which is the message content.
    #' @param func A Callback function.
    on_message = function(func) {
      self$ws$onMessage(func)
    },
    #' @description Called when the connection fails to be established.
    #' The event will have an 'message' element, a character vector of length 1
    #' describing the reason for the error.
    #' @param func A Callback function.
    on_error = function(func) {
      self$ws$onError(func)
    },
    #' @description Send a message to the server.
    #' @param msg Messages.
    send = function(msg) {
      self$ws$send(msg)
    },
    #' @description Close the connection.
    close = function() {
      self$ws$close()
    }
  )
)
